### PR TITLE
libc/minimal: Add rand and srand

### PR DESCRIPTION
**Pull request by yashi**
rand() and srand() are pseudo-random number generator functions
defined in ISO C.  This implementation uses the Linear congruential
generator algorithm with the following parameters, which are the same
as GNU Libc.

  Modulus 0x7fffffff
  Multiplier 1103515245
  Increment 12345

ISO C / POSIX.1-2001 uses bits high 15 bits for Modulus, thus RAND_MAX
32767.  ref rand(3).

Signed-off-by: Yasushi SHOJI <yashi@spacecubics.com>

### CODE REVIEW

**Comment by yashi** [comment]
Or should I call `sys_rand32_get()`, instead?

**Comment by yashi** [comment]
@enjiamai, @KangJianX would mind to give me some feedback?

**Comment by enjiamai** [comment]
Hi @yashi ,  OK, please give me some time to take look at this, I'll feedback to you later. thanks!

**Comment by enjiamai** [comment]
Hi @yashi , I have no problem with this implementation, it looks good, and I think we need this. I would so appreciate it if some basic test cases for this can also be added to tests/lib/c_lib/src/. How do you think? thank you!

**Review by enjiamai — CHANGES_REQUESTED** [review]
Could you please help to add the test cases for these, thank you!

**Review by enjiamai — APPROVED** [review]
Looks great~ no problem from my side, thanks!

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [inline lib/libc/minimal/source/stdlib/rand.c:25]: What's the reason for defining underscore version of a function, and then add a weak alias with normal name? (Why not use the normal name directly?)

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [reply lib/libc/minimal/source/stdlib/rand.c:15]: Another question: given that the function is defined in terms of standard C types (int/unsigned), then why uint32_t is used here?

Comment (pfalcon) [reply lib/libc/minimal/source/stdlib/rand.c:15]: Final question: how rand() will behave, if srand() is not explicitly called, given such seed variable definition, and comment below `/* seed of 0 doesn't generate a good random number */` ?

File: tests/lib/c_lib/src/main.c
Comment (pfalcon) [inline tests/lib/c_lib/src/main.c:812]: These tests look good, but ideally, the test would verify that rand() results after calling srand() with the same value are reproducible (per the definition of srand() semantics). I.e.: call srand() with one value, and check that a few (2-3) rand() calls give some values, then call srand() with another value, and call rand() a few times and check that values are different from 1st time. Then call srand() again with the initial value, and check that rand() give the same values as first time.

(That's in addition to the existing testcases - again, they're fine.)

**Review by pfalcon — CHANGES_REQUESTED** [review]
I'm always +1 for sustainably growing the minimal library. But IMHO there're few cleanups to be done and questions to be answered before this is ready for inclusion.

Another note - we usually try to put tests into a separate commit vs the core changes.

**Comment by pfalcon** [comment]
I also find the current commit message to be confusing:

```
            initstate(0, state, sizeof(state));
            printf("%d\n", rand());

            return 0;
    }

See initstate(3p) for more detail.
```

You don't implement initstate(), so why people should see its man? I however followed the advice and what I see is:

```
       long int random(void);

       void srandom(unsigned int seed);

       char *initstate(unsigned int seed, char *state, size_t n);

       char *setstate(char *state);
```

In other words, rand/srand vs random/initstate seems to be two parallel, independent pseudo-random number generation function sets. You here don't implement the latter, so references to initstate() should rather be removed. Likewise, please double-check that rand/srand implementation matches that of glibc. If it is, please update the sample in the commit message to use rand/srand. And if it's not, I'd suggest to just use the sample implementation from POSIX: https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html and remove references to glibc.

(And if there's some confusion on my side, please let me know what I'm missing.)

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c:15]: for `uint32_t`, it was just my habit to use size specific type when ever possible.  I'll change it to `unsigned int` if required.

Because the seed is static, it's `0` if `srand()` is not called.  Thus, it returns `1103527590`.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c:25]: I followed other functions in minimal libc.  Why do they have aliases?

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c:15]: oops, `rand()` without calling `srand()` means that `rand()` will use `0` as the seed and returns `12345` instead of `1103527590` (which is the value for the seed `1`).

hmm... I'm not sure we should guard seed of 0 or not.  Does anyone have any idea?

**Comment by yashi** [comment]
In order to test the Linear congruential generator algorithm, which is refered as `TYPE_0` in GLIBC, you have to call `initstate()` with [a state array of size 8 bytes][1] or [smaller than 32 byte][2].  Please note that the man page you referred is `3` not [`3p`, or `3posix`, from Linux man-pages project][2].

I've updated my commit log as below for more clarity.
> See initstate(3p) for more detail about how to use LCG in GLIBC.

[1]: https://www.redhat.com/en/blog/understanding-random-number-generators-and-their-limitations-linux
[2]: https://man7.org/linux/man-pages/man3/initstate.3p.html

File: tests/lib/c_lib/src/main.c
Comment (yashi) [reply tests/lib/c_lib/src/main.c:812]: `zephyr/tests/lib/c_lib/README.txt` states:
> It is intended to catch issues in which a library is completely absent or non-functional, and is NOT intended to be a comprehensive test suite of all functionality provided by the libraries.

Regardless, I've added a reproducibility test for `rand()`.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [reply lib/libc/minimal/source/stdlib/rand.c:25]: > I followed other functions in minimal libc. Why do they have aliases?

Where do you see that?

```
zephyr/lib/libc/minimal/source$ grep -r -i weak *
zephyr/lib/libc/minimal/source$ grep -r -i alias *
```

That's exactly my point - existing sources in zephyr/lib/libc/minimal/source/ **don't** have any aliasing like that, and I'd suggest to not add it here either.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [reply lib/libc/minimal/source/stdlib/rand.c]: > Does anyone have any idea?

Yes, it should be something like:

```
 static unsigned long srand_seed = 1; 
```

(Fixing also other issues like double-underscore prefix; variable type is based on sample implementation in https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html)

File: tests/lib/c_lib/src/main.c
Comment (pfalcon) [reply tests/lib/c_lib/src/main.c:812]: Sure, that was just a suggestion, as long as you went to write a more or less detailed test at all. I appreciate adding that test.

**Comment by pfalcon** [comment]
@yashi 

> In order to test the Linear congruential generator algorithm, which is refered as TYPE_0 in GLIBC, you have to call initstate() with a state array of size 8 bytes or smaller than 32 byte.

That's an interesting implementation detail of GLIBC, but I don't see how that relates to the functions srand()/rand() which are being implemented here. Nothing in https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html mentions initstate(). Nor you implement initstate(). So, why do you mention it at all?

Let's start from the beginning. Do you want Zephyr's implementation of srand()/rand() to have the same behavior/produce the same values as Glibc's srand()/rand()? If so, I'm generally +1 for that, as Glibc is a well known library and I'm for being consistent with well-know de-facto standard. But then please sure that you exactly follow behavior of srand()/rand() and don't involve unrelated implementation details, like the fact that in Glibc, srand() is implemented in terms of initstats().

And well, if you don't have that specific requirement to follow Glibc behavior, just use the simple algorithm from "Generating the Same Sequence on Different Machines" subheading of https://pubs.opengroup.org/onlinepubs/9699919799/functions/rand.html .

> Please note that the man page you referred is 3 not 3p, or 3posix, from Linux man-pages project.

That page still described the functions:
```
       char *initstate(unsigned seed, char *state, size_t size);
       long random(void);
       char *setstate(char *state);
       void srandom(unsigned seed);
```

And not srand()/rand().

**Comment by yashi** [comment]
@pfalcon, thanks for your feedback.  The reason why I've chosen GLIBC LCG is that I assumed most of the developers in the Zephyr community prefer to have the same algorithm as their development environment, which is, I also assume, Linux with GLIBC.  I don't have any preferences at all.  And the reason I put the reference in the commit log is that when someone wants to verify my implementation, it'd be nice to give a point or two for them.  I thought it'd be helpful for future references.  Nothing else.

I followed GLIBC LCG's _behavior_ in the Zephyr implementation.  No detail, nor `TYPE_0` thing is in Zephyr implementation.  The reference is just for a hint who wants verification in behavior.  If the default algorithm in GLIBC is LCG, I didn't put the reference in the commit log, but it's not.  If you just run `rand()` and `srand()` on your Linux, you are using a different algorithm and you get different values.  I'm not sure why the reference/doc in the commit log is such a big issue.

Again, I don't have any preferences in the algorithms.  I just want `rand()` and `srand()` for compatibility.  I can either remove the reference from the commit log or implement the C standard.  Please let me know which way you want.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c]: Oops! Thanks.  I was checking `lib/libc/newlib/libc-hooks.c`.  Fixed.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c]: fixed.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [reply lib/libc/minimal/source/stdlib/rand.c]: > Oops! Thanks. I was checking lib/libc/newlib/libc-hooks.c. Fixed.

I see. That's actually for newlib libc. Thanks for update!

File: lib/libc/minimal/source/stdlib/rand.c
Comment (pfalcon) [inline lib/libc/minimal/source/stdlib/rand.c:9]: Is this include really needed?

**Comment by pfalcon** [comment]
@yashi 

> The reason why I've chosen GLIBC LCG is that I assumed most of the developers in the Zephyr community prefer to have the same algorithm as their development environment, which is, I also assume, Linux with GLIBC.

I see, understand and appreciate such an approach. My concern is to make sure that if this approach is taken, then it actually works consistently, and described in a way to avoid confusion.

I now played with the sources myself, and took a closer look at glibc source (ref: https://code.woboq.org/userspace/glibc/stdlib/random.c.html#unsafe_state), and understood the situation: the default algorithm in glibc is "TYPE_3", while you implement "TYPE_0" here. Thus, to get compatible output from glibc, we need to switch it to TYPE using that initstate() call, after which both srand() and rand() will work in a way compatible with the implementation here.

I thus would recommend to update the commit message to explain that, e.g.:

> This implementation uses the Linear Congruential
Generator (LCG) algorithm with the following parameters, which are the
same as **used in** GNU Libc **"TYPE_0" algorithm**. **Note that the default
algorithm used by GNU Libc is not TYPE_0, and TYPE_0 should be selected
first by an initstate() call as shown below.**

> ...

    #include <stdio.h>
    #include <stdlib.h>

    int main()
    {
            static char state[8];

            /* Switch GLIBC to use LCG/TYPE_0 generator type. */
            initstate(0, state, sizeof(state));

            srand(1);  /* Or any other value. */
            printf("%d\n", rand());
            printf("%d\n", rand());

            return 0;
    }

**Review by pfalcon — APPROVED** [review]
I'd recommend to clarify situation re: GLIBC compatibility, as discussed in the previous comment.

Otherwise, looks good.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c:9]: nop.  I'll leave `stdlib.h` there for checking the `rand()` and `srand()` declarations against the definitions.

**Comment by yashi** [comment]
It seems like I need to setup memory domain for static storage, `srand_seed`, to pass the tests from `tests/crypto/mbedtls`.  For newlib, [as documented here][1], there seems to be `z_libc_partition` but I'm not sure I can use it for minimal libc.

Would you please give me some hint how I should go with it?

[1]: https://docs.zephyrproject.org/latest/reference/usermode/memory_domain.html#memory-domains

**Comment by yashi** [comment]
spent a few hours reading memory partitions and memory domains, and came up with the updated patch set. The following twister command gives me the same result as the current master 0c4e9d29bbc1b.  (all passed except for failed native_posix and skipped qemu_x86_tiny w/ both SDK 0.12.4 and 0.13.0 alpha 1).
```
$ scripts/twister -N -T tests/crypto/mbedtls -v
```
Hope I got it right.

**Review by tejlmand — APPROVED** [review]

**Comment by pfalcon** [comment]
@yashi, Thanks for looking further into this. The situation reminds me my experience with hacking POSIX and/or libc subsystems, when seemingly simple changes leed to cascade of issues and follow-up changes. So, you're not alone ;-).

I looked into the situation myself. mps2_an385 supports 8 MPU regions, and one after another protection features enabled, that trickles into 3 "free" partition slots available for application code. Having additional libc partition indeed gets this down to 2 free partition slots, which causes the mem_protect/userspace test to fail.

I was able to get it pass with `CONFIG_TEST_HW_STACK_PROTECTION=n`. So, we could potentially make another test config for mps2_an385 with that option and exclude it from normal tests. I don't think that would scale, because 8 MPU regions is standard for Cortex-M0/Cortex-M3 devices, so the same problem would happen to many other (real) devices, the only special thing about mps2_an385 is that it's emulated via qemu, and we run this emulation during CI. Try to filter all affected devices would be too adhoc and cumbersome, and I guess we'd really like to run the rest with stack protection anyway.

So, I'd suggest to do partition plucking on where we add it now - on minimal libc size. It shouldn't be too bread-crumb configuration either, so instead of e.g. CONFIG_MINIMAL_LIBC_RAND, I'd suggest to add something like CONFIG_MINIMAL_LIBC_GLOBALS with a description like "Support C library functions which implicitly access global/static data, e.g. rand(). Such data must be put into a memory partition, and disabling this option may save an entry for application defining many custom partitions. ..."

Then disable this option for the affected test(s).

**Comment by yashi** [comment]
@pfalcon, Thank you for your suggestion.  I've implemented what you suggested (in a way I understand, at least).   Please take a look the commit "libc: minimal: Use a memory partition iff globals are required".
Any feedback is as always welcome. :smiley:

**Comment by pfalcon** [comment]
> Please take a look the commit "libc: minimal: Use a memory partition iff globals are required".
Any feedback is as always welcome.

Yes, that looks pretty much how I pictured it in my mind, but I'd suggest to not change  tests/kernel/mem_protect/userspace/testcase.yaml to add another test variant, bit instead just add "CONFIG_MINIMAL_LIBC_GLOBALS=n" to its prj.conf. After all, the whole idea of introducing CONFIG_MINIMAL_LIBC_GLOBALS was to avoid adding yet another test variant for mem_protect/userspace (which already has a few of them).

So, my suggestion would be to add to tests/kernel/mem_protect/userspace/prj.conf following:

```
# Disable libc globals partition to make the test runnable on devices
# with 8 MPU regions.
CONFIG_MINIMAL_LIBC_GLOBALS=n
```

**Comment by pfalcon** [comment]
And another suggestion is to squash related commits together, e.g.:
* libc: minimal: Add rand() and srand()
* libc: minimal: Make srand_seed memory partition aware
* libc: minimal: Use a memory partition iff globals are required 

should be one; 2 of "tests: c_lib:" commits can also be squashed together.

**Comment by pfalcon** [comment]
> Just noticed that in #26545, it's using LIBC_PARTITION instead of LIBC_PART as in this PR.

I guess LIBC_PART is fine. And that PR shows that it may be not easy to get bigger changes thru, as they may "pull strings" for even more changes, and some PRs get stale. So again, thanks, thanks for looking into related issues here, hopefully, it will make it easier to add other libc's/functions later.

**Comment by yashi** [comment]
@pfalcon, I've updated the patchset.  Fixed the prj.conf instead of the testcase, merged "libc: minimal:" commits in one (merged all 4 because "libc: minimal: Place globals in z_libc_partition" causes a regression if you test it alone), merged two tests in one.

I'd like to merge "tests: crypto: mbedtls: Remove local implementation of rand()" into "libc: minimal: Add rand() and srand()" because without the fix, test for mbedtls failed to build.  WDYT?

**Comment by pfalcon** [comment]
@yashi 

> I'd also like to sqiash "tests: crypto: mbedtls: Remove local implementation of rand()" into "libc: minimal: Add rand() and srand()" because without the fix, tests for mbedtls failed to build. WDYT?

Well, I'd personally err on the side of having clearly delinated commits rather than on having perfect bisectability (IMHO it's better to have clean commits and 98% bisectability rather than unwieldy commits and 100% bisectability). But you can see if there will be other review comments before it's merged.

Thanks for all the updates. As it stands, this PR has substantial changes and apparently doesn't fit with codefreeze of the 2.6 release in progress, so it'll be merged when the merge window opens.

Then, we'll need to remind mergers that build error of

```
Error: lib/libc/minimal/include/stdlib.h:45:WARNING: Violation to rule 21.2 (Should not used a reserved identifier) - srand
```

is false positive here - this patch **does** implement that function, as part of libc.

**Comment by yashi** [comment]
@pfalcon, OK, let's see what the others have to say.  Meanwhile, I've updated a commit log by adding the false positive.

> this PR has substantial changes and apparently doesn't fit with codefreeze of the 2.6 release in progress, so it'll be merged when the merge window opens.

Make sense.  :+1:

Thank you, @pfalcon, for all your help and advice.  I really appreciate it. :pray:

**Comment by yashi** [comment]
can we merge this?

**Comment by yashi** [comment]
@stephanosio, I saw the MAINTAINERS.yml change and you are now the maintainer of libcs.  Would you mind reviewing this PR? I just rebased to the main. Thanks.

File: lib/libc/Kconfig
Comment (stephanosio) [inline lib/libc/Kconfig:155]: Noting that we already have the configs to selectively enable/disable different features of the minimal libc, I think it would be better to add a config to enable the "rand feature" and make that select `NEED_LIBC_MEM_PARTITION` so that the globals can be used for userspace.

For example:
```
config MINIMAL_LIBC_RAND
	bool "Enable minimal libc rand implementation"
	select NEED_LIBC_MEM_PARTITION
...
```

File: include/sys/libc-hooks.h
Comment (stephanosio) [inline include/sys/libc-hooks.h:71]: `CONFIG_MINIMAL_LIBC_GLOBALS` should not be necessary if we use `CONFIG_NEED_LIBC_MEM_PARTITION` as recommended above.

**Review by stephanosio — CHANGES_REQUESTED** [review]

File: lib/libc/Kconfig
Comment (stephanosio) [reply lib/libc/Kconfig:155]: p.s. I see that the PR originally added `MINIMAL_LIBC_RAND`, but later changed to `MINIMAL_LIBC_GLOBALS` because memory protection tests were failing due to the libc memory partition being added and there being insufficient number of MPU regions being available for low end Cortex-M cores.

FWIW, I do not think it warrants reworking the config scheme just to get around that.

The current minimal libc implementation explicitly does not make use of the globals due to the high cost associated with it (i.e. additional memory partition for libc). Noting that, I think it is better to just make any minimal libc features that `select NEED_LIBC_MEM_PARTITION` not enabled by default, so users can enable those features if and only if they need them.

File: lib/libc/minimal/CMakeLists.txt
Comment (stephanosio) [inline lib/libc/minimal/CMakeLists.txt:14]: Need to remove this as per the `zephyr_library_sources_ifdef` below.

**Review by stephanosio — CHANGES_REQUESTED** [review]

**Comment by cfriedt** [comment]
@yashi - can you please address any remaining comments?

**Comment by yashi** [comment]
Yes. Will do it in this weekend.

File: lib/libc/Kconfig
Comment (yashi) [reply lib/libc/Kconfig]: Reverted back to `CONFIG_MINIMAL_LIBC_RAND` instead of `CONFIG_MINIMAL_LIBC_GLOBALS`.

Comment (yashi) [reply lib/libc/Kconfig]: And also `CONFIG_MINIMAL_LIBC_RAND` now selects `CONFIG_NEED_LIBC_MEM_PARTITION`.

File: lib/libc/minimal/CMakeLists.txt
Comment (yashi) [reply lib/libc/minimal/CMakeLists.txt]: done

File: CMakeLists.txt
Comment (yashi) [reply CMakeLists.txt:997]: I don't know what I was doing.  But indeed it works without it.  Reverted.

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [inline lib/libc/minimal/source/stdlib/rand.c:11]: @stephanosio, is this the right way to specify the global variable in minimal libc?

File: include/sys/libc-hooks.h
Comment (yashi) [reply include/sys/libc-hooks.h]: done

File: lib/libc/minimal/source/stdlib/rand.c
Comment (stephanosio) [reply lib/libc/minimal/source/stdlib/rand.c:11]: Yes. `z_libc_partition` is valid as long as `CONFIG_NEED_LIBC_MEM_PARTITION=y`.

File: doc/guides/porting/arch.rst
Comment (gmarull) [inline doc/guides/porting/arch.rst:798]: nit: use `:kconfig:` to reference Kconfig options

File: include/sys/libc-hooks.h
Comment (stephanosio) [inline include/sys/libc-hooks.h:78]: I am not sure if I understand this comment.

The scheme post-this PR is as simple as:
1. All newlib globals will be placed into `z_libc_partition`.
2. Minimal C library globals, if any, will be placed into `z_libc_partition`.
2. Stack canary globals will be placed into `z_libc_partition` since it is not worth placing in its own partition.
3. Some architectures may place the global pointer to the thread local storage in `z_libc_partition` since it is not worth placing in its own partition.

File: lib/libc/Kconfig
Comment (stephanosio) [inline lib/libc/Kconfig:150]: Should not be `default y` as per https://github.com/zephyrproject-rtos/zephyr/pull/32667#discussion_r656414206.

Any **minimal** C library features that predicate on such a heavy requirement as requiring an additional memory partition should not be enabled by default, and should only be enabled on demand.

**Review by stephanosio — CHANGES_REQUESTED** [review]
Some additional comments

File: lib/libc/minimal/source/stdlib/rand.c
Comment (yashi) [reply lib/libc/minimal/source/stdlib/rand.c:11]: Thanks.

File: doc/guides/porting/arch.rst
Comment (yashi) [reply doc/guides/porting/arch.rst:798]: done.  Thanks.

Comment (yashi) [reply doc/guides/porting/arch.rst:798]: done.  Thanks.

File: doc/reference/usermode/memory_domain.rst
Comment (yashi) [reply doc/reference/usermode/memory_domain.rst:279]: done.  Thanks.

File: lib/libc/Kconfig
Comment (yashi) [reply lib/libc/Kconfig:150]: done.  Thanks.

File: include/sys/libc-hooks.h
Comment (yashi) [reply include/sys/libc-hooks.h:78]: I just copy & pasted your comment. Hope you don't mind.

**Comment by cfriedt** [comment]
fixes #6493

**Review by stephanosio — APPROVED** [review]
LGTM

**Pull request merged**
